### PR TITLE
PBR Update, Environment.

### PR DESCRIPTION
**Pull request by sebavan**
Early PR for review, still on it tomorrow, please, do not merge yet.

### CODE REVIEW

**Comment by sebavan** [comment]
Fixes #2368 and #2080

File: src/Materials/PBR/babylon.pbrBaseMaterial.ts [hunk 1]
```diff
@@ -679,31 +689,35 @@
...
-                        if (this._refractionTexture instanceof HDRCubeTexture) {
-                            defines.REFRACTIONMAPINLINEARSPACE = true;
+                    }
 
-                            if ((<HDRCubeTexture>this._refractionTexture).isPMREM) {
-                                defines.USEPMREMREFRACTION = true;
-                            }
+                    if (this._environmentBRDFTexture && StandardMaterial.ReflectionTextureEnabled) {
+                        // This is blocking.
+                        if (!this._environmentBRDFTexture.isReady()) {
```
Comment (deltakosh) [inline src/Materials/PBR/babylon.pbrBaseMaterial.ts]: not convince by the name..Could we make it more explicit? Should I know what BRDF to use the material?

File: src/Materials/PBR/babylon.pbrBaseMaterial.ts [hunk 2]
```diff
@@ -1063,22 +1077,21 @@
                             this._uniformBuffer.updateMatrix("bumpMatrix", this._bumpTexture.getTextureMatrix());
                         }
 
-                        if (this._refractionTexture && StandardMaterial.RefractionTextureEnabled) {
-                            this._microsurfaceTextureLods.y = Math.round(Math.log(this._refractionTexture.getSize().width) * Math.LOG2E);
-
+                        var refractionTexture = this._getReractionTexture();
```
Comment (deltakosh) [inline src/Materials/PBR/babylon.pbrBaseMaterial.ts]: getre**F**raction

File: src/Materials/Textures/babylon.baseTexture.ts [hunk 3]
```diff
@@ -228,6 +251,46 @@
             }
         }
 
+        public get sphericalPolynomial(): SphericalPolynomial {
+            if (!this._texture || !Internals.CubeMapToSphericalPolynomialTools || !this.isReady()) {
```
Comment (deltakosh) [inline src/Materials/Textures/babylon.baseTexture.ts]: should add isCube as well

File: src/Math/babylon.math.ts [hunk 4]
```diff
@@ -5108,6 +5115,64 @@
             this.L21 = this.L21.scale(scale);
             this.L22 = this.L22.scale(scale);
         }
+
```
Comment (deltakosh) [inline src/Math/babylon.math.ts]: can we extract this class to a file?

File: src/Shaders/ShadersInclude/pbrFunctions.fx [hunk 5]
```diff
@@ -130,4 +106,37 @@ float computeDefaultMicroSurface(float microSurface, vec3 reflectivityColor)
 float fresnelGrazingReflectance(float reflectance0) {
 	float reflectance90 = clamp(reflectance0 * 25.0, 0.0, 1.0);
 	return reflectance90;
+}
+
+// To enable 8 bit textures to be used we need to pack and unpack the LOD
+// Inverse alpha is used to work around low-alpha bugs in Edge and Firefox
+#define UNPACK_LOD(x) (x)
```
Comment (deltakosh) [inline src/Shaders/ShadersInclude/pbrFunctions.fx]: useful?

File: src/babylon.engine.ts [hunk 6]
```diff
@@ -2465,7 +2462,7 @@
 
                          var loadMipmap = (info.isRGB || info.isLuminance || info.mipmapCount > 1) && !noMipmap && ((info.width >> (info.mipmapCount - 1)) === 1);
                          prepareWebGLTexture(texture, this._gl, scene, info.width, info.height, invertY, !loadMipmap, info.isFourCC, () => {
-                        //     Internals.DDSTools.UploadDDSLevels(this, data, info, loadMipmap, 1);
+                             Internals.DDSTools.UploadDDSLevels(this, data, info, loadMipmap, 1);
```
Comment (deltakosh) [inline src/babylon.engine.ts]: stupid me

File: src/babylon.engine.ts [hunk 7]
```diff
@@ -3117,7 +3114,79 @@
...
+                    return;
+                }
+
+                const textures: BaseTexture[] = [];
+                for (let i = 0; i < mipSlices; i++) {
+                    //compute LOD from even spacing in smoothness (matching shader calculation)
+                    let smoothness = i / (mipSlices - 1);
+                    let roughness = 1 - smoothness;
+                    const kMinimumVariance = 0.0005;
+                    let alphaG = roughness * roughness + kMinimumVariance;
```
Comment (deltakosh) [inline src/babylon.engine.ts]: kMinimumVariance ==> minimumVariance

File: src/babylon.engine.ts [hunk 8]
```diff
@@ -3117,7 +3114,79 @@
...
+                    var glTextureFromLod = gl.createTexture();
+                    glTextureFromLod.isCube = true;
+                    this._bindTextureDirectly(gl.TEXTURE_CUBE_MAP, glTextureFromLod);
+
+                    gl.texParameteri(gl.TEXTURE_CUBE_MAP, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
+                    gl.texParameteri(gl.TEXTURE_CUBE_MAP, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
+                    gl.texParameteri(gl.TEXTURE_CUBE_MAP, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
+                    gl.texParameteri(gl.TEXTURE_CUBE_MAP, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
+
+                    if (loadData.isDDS) {
```
Comment (deltakosh) [inline src/babylon.engine.ts]: this is always true

File: src/babylon.engine.ts [hunk 9]
```diff
@@ -3527,6 +3578,17 @@
             if (index !== -1) {
                 this._loadedTexturesCache.splice(index, 1);
             }
+
+            // Intergated fixed lod samplers.
```
Comment (deltakosh) [inline src/babylon.engine.ts]: typo :)

File: src/babylon.engine.ts [hunk 10]
```diff
@@ -3969,40 +4031,39 @@
...
             if (faceIndex > -1) {
-                gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_CUBE_MAP_POSITIVE_X + faceIndex, texture, lodIndex);            
+                gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_CUBE_MAP_POSITIVE_X + faceIndex, texture, 0);
             } else {
-                gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, texture, lodIndex);
+                gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, texture, 0);
             }
 
-            let readFormat = (texture.type !== undefined) ? this._getRGBABufferInternalSizedFormat(texture.type) : gl.RGBA;
+            let readFormat = (texture.format !== undefined) ? this._getRGBABufferInternalSizedFormat(texture.format) : gl.RGBA;
```
Comment (deltakosh) [inline src/babylon.engine.ts]: unused

File: src/Materials/Textures/babylon.baseTexture.ts [hunk 11]
```diff
@@ -228,6 +251,46 @@
...
+            return this._texture._sphericalPolynomial;
+        }
+
+        public set sphericalPolynomial(value: SphericalPolynomial) {
+            if (this._texture) {
+                this._texture._sphericalPolynomial = value;
+            }
+        }
+
+        public get lodTextureHigh(): BaseTexture {
```
Comment (deltakosh) [inline src/Materials/Textures/babylon.baseTexture.ts]: we should flag these functions as internal with _

File: src/Materials/Textures/babylon.cubeTexture.ts [hunk 12]
```diff
@@ -8,19 +8,25 @@
         private _extensions: string[];
         private _textureMatrix: Matrix;
         private _format: number;
+        private _prefiletered: boolean;
```
Comment (deltakosh) [inline src/Materials/Textures/babylon.cubeTexture.ts]: typo

File: src/Materials/Textures/babylon.hdrCubeTexture.ts [hunk 13]
```diff
@@ -484,88 +488,9 @@ module BABYLON {
...
-
-                // Callback.
-                callback(buffer);
-            }
-
-            // Download and process.
-            Tools.LoadFile(url, data => {
-                getDataCallback(data);
-            }, null, null, true, onError);
+            // Coming Back in 3.1.
```
Comment (deltakosh) [inline src/Materials/Textures/babylon.hdrCubeTexture.ts]: lol a teaser :D

File: src/Shaders/pbr.fragment.fx [hunk 14]
```diff
@@ -659,4 +755,12 @@ void main(void) {
 	//// Emissive Color
 	//vec2 test = vEmissiveUV * 0.5 + 0.5;
 	//gl_FragColor = vec4(test.x, test.y, 1.0, 1.0);
+
```
Comment (deltakosh) [inline src/Shaders/pbr.fragment.fx]: comments can be removed :)

File: src/babylon.engine.ts (hunk 8 above)
Comment (deltakosh) [inline src/babylon.engine.ts]: Or we need to add more info the the onLoad in createCubeTexture
if we expect only DDS (which is my understanding): we should either add it in the comment or be more explicit in a console.warn

File: src/Materials/Textures/babylon.baseTexture.ts (hunk 3 above)
Comment (sebavan) [inline src/Materials/Textures/babylon.baseTexture.ts]: I keep it for planar, the check is currently in the tool.

File: src/Materials/PBR/babylon.pbrBaseMaterial.ts (hunk 1 above)
Comment (sebavan) [inline src/Materials/PBR/babylon.pbrBaseMaterial.ts]: It is called like this in the nomenclature I found and I am really bad at naming :-)

File: src/babylon.engine.ts (hunk 8 above)
Comment (sebavan) [inline src/babylon.engine.ts]: I am expecting to add more support here for ktx for instance, I ll add a warning.

File: src/Shaders/pbr.fragment.fx (hunk 14 above)
Comment (sebavan) [inline src/Shaders/pbr.fragment.fx]: There are debug switch :-) just uncomment and you see the channel, I ll remove once Spector supports it.

**Pull request merged**